pkg/utils/retry/retry.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package retry
+
+import (
+	"time"
+)
+
+// Params are used to define how retry attempts are handled.
+type Params struct {
+	MaxRetries     uint
+	InitialBackoff time.Duration
+	BackoffFactor  float64
+}
+
+// Invocation represents a function that is desired to be retried until it succeeds (i.e. it returns nil).
+type Invocation func() error
+
+// Retry retries the given Invocation based on the given Params until it returns no error, at which point this
+// function returns no error as well.
+// If the retry attempts are exhausted, this function returns the most recent error returned from the given Invocation.
+func Retry(invocation Invocation, params *Params) error {
+	var err error
+
+	backoff := params.InitialBackoff
+
+	// There is always a mandatory initial attempt at running funcToRetry.
+	// The delays only start if this attempt failed and params.MaxRetries is at least 1.
+	for retries := uint(0); retries <= params.MaxRetries; retries++ {
+		if retries != 0 {
+			time.Sleep(backoff)
+		}
+
+		err = invocation()
+		if err == nil {
+			return nil
+		}
+
+		if retries != 0 && retries < params.MaxRetries {
+			backoff = time.Duration(float64(backoff) * params.BackoffFactor)
+		}
+	}
```

Comment:
invocation function shoud be executed before sleep.
Also, `MaxRetries` should be `int64` otherwise the example below will not work.
```
	for retries := params.MaxRetries; retries >= 0; retries-- {
		// first, we need to execute invocation function
		if err = invocation(); err == nil {
			return nil
		}
		// than  sleep
		time.Sleep(backoff)
		// than  calculate a new backoff duration
		backoff = time.Duration(float64(backoff) * params.BackoffFactor)
	}
```

Reply:
@soluchok I don't think that's right. If I execute the invocation before the sleep then for the first retry (`retries`=1), it won't delay at all since I have the `if retries != 0` check. Also, it'll do an unnecessary sleep after the final attempt.

I don't understand your comment about changing `MaxRetries` to be an `int64`. I'm not sure what your example is supposed to illustrate - are you suggesting I change the loop to look like that? Also, in that example loop the time.Sleep on the last retry is unnecessary.

Reply:
sorry, my fault :)